Move timestamp bindings in STORAGE

petrosagg opened the pull request:
### Motivation

This PR moves timestamp binding storage out of the coordinator and into the storage controller. The timestamp binding on-disk representation backed by a `Stash` instance. The intent is to replace `Stash` with a `persist` collection as soon as we are comfortable with that.

The timestamp binding feedback coming from workers is not a differential stream, it resembles more of an upsert stream where every statement invalidates the previous one. Previously we were storing the data in SQLite with the same semantics. In order to store in differential collection in `Stash` in a way that can play well with the standard collection compaction semantics the STORAGE controller differentializes all bindings received before storing them and "integrates" them before sending them down to workers.

There is some work left for the future of changing the workers to also work with a differential collection of bindings, just as if they received them from a `persist` collection directly, but that was too much of a change for a single PR.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

benesch commented on src/dataflow-types/src/client.rs:
```diff
@@ -441,6 +443,8 @@ impl<'a, C: Client + 'a> futures::stream::Stream for SelectStream<'a, C> {
             let result = {
                 use futures::Future;
+                // TODO(petrosagg): this implementation seems incorrect as on every poll a new recv Future
+                // is constructed, polled once, and then dropped. Investigate
```
Eep!

benesch commented on src/dataflow-types/src/client/controller/storage.rs:
```diff
@@ -264,6 +303,80 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
+            let since = ts_binding_stash.since().unwrap();
+            let mut seal_frontier = upper.clone();
+
+            // Here we differentialize the bindings we got from workers
```
What would be really helpful for me, at least, is to describe this differentialization in prose! I've totally paged out the details of that conversation that we had long ago.

benesch reviewed (commented):
Neat! This pretty much all makes sense to me. I think @frankmcsherry may have more opinions.

There's a lot of `unwraps()` around. What is our plan for those? I think in many cases we just need to exit if we can't persist timestamp bindings, but would be nice to handle those a bit more gracefully than panicking.

frankmcsherry commented on src/dataflow-types/src/client/controller.rs:
```diff
@@ -39,6 +40,11 @@ mod storage;
 pub use compute::{ComputeController, ComputeControllerMut};
 mod compute;
 
+pub struct ControllerConfig<C> {
+    pub dataflow_client: C,
+    pub state_dir: PathBuf,
+}
```
Doc comments please!

frankmcsherry commented on src/dataflow-types/src/client/controller/storage.rs:
```diff
@@ -105,35 +123,68 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
     pub async fn create_sources(
         &mut self,
-        mut bindings: Vec<CreateSourceCommand<T>>,
+        mut bindings: Vec<(GlobalId, (crate::sources::SourceDesc, Antichain<T>))>,
```
Is it easy to explain this change?

frankmcsherry commented on src/dataflow-types/src/client/controller/storage.rs:
```diff
@@ -169,23 +221,6 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
             .await
             .map_err(StorageError::from)
     }
-    /// Downgrade the read capabilities of specific identifiers to specific frontiers.
-    ///
-    /// Downgrading any read capability to the empty frontier will drop the item and eventually reclaim its resources.
-    async fn allow_compaction(
-        &mut self,
-        frontiers: Vec<(GlobalId, Antichain<T>)>,
-    ) -> Result<(), StorageError> {
-        // Validate that the ids exist.
-        self.as_ref()
-            .validate_ids(frontiers.iter().map(|(id, _)| *id))?;
-
-        let policies = frontiers
-            .into_iter()
-            .map(|(id, frontier)| (id, ReadPolicy::ValidFrom(frontier)));
-        self.set_read_policy(policies.collect()).await;
-        Ok(())
-    }
```
This is .. unrelated refactoring or ?

frankmcsherry commented on src/dataflow-types/src/client/controller/storage.rs:
```diff
@@ -252,9 +287,13 @@ impl<'a, T: Timestamp + Lattice> StorageController<'a, T> {
-    pub(super) fn collection_mut(
+impl<'a, C: Client<T>, T> StorageControllerMut<'a, C, T>
+where
+    T: Timestamp + Lattice + TryInto<i64> + TryFrom<i64>,
+    <T as TryInto<i64>>::Error: std::fmt::Debug,
+    <T as TryFrom<i64>>::Error: std::fmt::Debug,
```
These constraints seem to be everywhere, and it would be great to localize them instead! For example, having them only on the `persist_timestamp_bindings` method, or better yet not at all and having the stash trait link them up only at the constructor time.

frankmcsherry commented on src/dataflow-types/src/client/controller/storage.rs:
```diff
@@ -105,35 +123,68 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
-        for binding in bindings.iter() {
-            let collection = CollectionState::new(binding.desc.clone(), binding.since.clone());
-            self.storage.collections.insert(binding.id, collection);
+        for (id, (desc, since)) in bindings {
```
With the vector of bindings now being returned, I think it is more clear that we aren't looking at `since` when we do the work here. Like, we push a bunch of binding information that is potentially irrelevant from `since`s point of view. I'm not sure it is harmful as I'm not certain what the vector communicates, but wanted to point it out! (I think previously we .. validated nothing).

frankmcsherry commented on src/dataflow-types/src/client/controller/storage.rs:
```diff
@@ -264,6 +303,80 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
+
+            // Here we differentialize the bindings we got from workers
+            ts_binding_stash
+                .update_many(bindings.iter().cloned().flat_map(|(pid, ts, offset)| {
```
I would either defensively sort `bindings` here, or put some ALLCAPS comments to the effect that it should be sorted by `ts`. It looks like that is necessary for correctess, but ignore if I'm wrong.

petrosagg replied:
no, stash doesn't require the operator to be sorted

frankmcsherry commented on src/dataflow-types/src/client/controller/storage.rs:
```diff
@@ -264,6 +303,80 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
+                    let update = ((pid, ()), ts, offset.offset - prev_offset.offset);
+
+                    if seal_frontier.less_than(&ts) {
+                        seal_frontier = Max1Antichain::Singular(ts);
```
The whole `Max1Antichain` thing continues to confuse me. It looks like `Stash` made that call, but .. yeah.

benesch replied:
What would you expect to see here with a normal antichain? Can we add that same API to `Max1Antichain`?

frankmcsherry replied:
That's what would be clearest. It's possible that I don't have the `Stash` requirements top of mind, but to the extent that `Max1Antichain` is a perf optimization, meh. It seemed like if it was an "I can be converted to `{ number, NULL }`" then it seemed that it could be a wrapper around `Antichain` with a `T: TotalOrder` constraint.

Not sure, but my opinions on "roll your own" TD/DD stuff has evolved. :D No rush to change things that work, but I end up needing to re-understand if it does what I think when it is a new type to me.

benesch replied:
I suppose it's sort of a perf optimization, but yeah, I'd phrase the requirement as restricting the antichain to a type we know we can represent in SQL. 

> then it could be a wrapper around Antichain with a T: TotalOrder constraint.

That sounds right! I think the Frank that reviewed the stash PR hadn't come up with that yet. :D

frankmcsherry commented on src/dataflow-types/src/client/controller/compute.rs:
```diff
@@ -123,7 +123,12 @@ impl<'a, T: Timestamp + Lattice> ComputeController<'a, T> {
 }
 
-impl<'a, C: Client<T>, T: Timestamp + Lattice> ComputeControllerMut<'a, C, T> {
+impl<'a, C: Client<T>, T> ComputeControllerMut<'a, C, T>
+where
+    T: Timestamp + Lattice + TryInto<i64> + TryFrom<i64>,
+    <T as TryInto<i64>>::Error: std::fmt::Debug,
+    <T as TryFrom<i64>>::Error: std::fmt::Debug,
+{
```
Mentioned elsewhere, but these constraints seem like noise. I think if you put them on an `impl` block for the (I believe) one method that requires them, then no one else should need them.

frankmcsherry reviewed (approved):
Looks good, thanks. I have a few comments about adding some docs and comments, and moving / subtracting type constraints around, but morally this seems good.

benesch commented on src/dataflow-types/src/client/controller/storage.rs:
```diff
@@ -264,6 +303,80 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
+
+            ts_binding_stash.seal(seal_frontier).unwrap();
+            ts_binding_stash.compact(compaction_frontier).unwrap();
+            ts_binding_stash.consolidate().unwrap();
```
Oh I meant to ask: do we need to do something smarter than consolidating the timestamp bindings on every update?

guswynn commented:
Would it be reasonable to file a TODO task to remove the `blocking` io happening within timely works (and I suppose, the `coord` thread as well)

benesch commented:
Where does the blocking on timely worker threads happen? It seems to only block the coord thread to me! The rule about blocking in an async task doesn't really apply here, since it has a dedicated thread, so not worried about that. It does impose a latency tax on the rest of the system, though, which is probably worth an issue.

aljoscha commented on src/expr/src/id.rs:
```diff
@@ -131,12 +131,37 @@ impl fmt::Display for SourceInstanceId {
 }
 
+use bytes::{Buf, BufMut};
+impl mz_persist_types::Codec for PartitionId {
```
All the other source struct for which we added `Codec` impls have protobuf-based impls. For example, take a look at `SourceTimestamp`.

guswynn commented:
@benesch oh I see, `StorageControllerMut` methods are only called on the coordinator thread?
definitely we want a followup to resolve this latency bottleneck, the entire system could wait on sqlite io

also, will this pr need a migration to read the old bindings out of the catalog?

benesch commented:
> @benesch oh I see, `StorageControllerMut` methods are only called on the coordinator thread? 

Yep!

> also, will this pr need a migration to read the old bindings out of the catalog?

Yeah, @petrosagg said he was working on them!

petrosagg commented on misc/python/materialize/mzcompose/services.py:
```diff
@@ -577,6 +578,11 @@ def __init__(
+            entrypoint.append(
+                "--validate-timestamp-bindings=/share/mzdata/catalog/controller/timestamp_bindings"
+            )
+
```
changed the flat to that

petrosagg commented on src/testdrive/src/action/verify_timestamp_compaction.rs:
```diff
@@ -70,7 +75,22 @@ impl Action for VerifyTimestampCompactionAction {
+                                    MzOffset { offset },
+                                )
+                            })
+                            .collect();
```
yep, changed

petrosagg commented:
Alright, this is finally ready for a final review. Some notes here for reviewers:

* @benesch all `.unwrap()`ed errors generated within Storage controller are all wired up to Result return types. Since the compute controller calls some of the now fallible methods all the transitive dependants are also changed to return a Result if they didn't already. The Result wrapping stops at the coordinator where all the calls to the controllers are `.unwrap()`ed. This is what we were doing already and I didn't want to delay this PR with further error type refactoring.
* @aljoscha the `Codec` implementation for `PartitionId` is now based on a protobuf schema
* @frankmcsherry I wasn't able to contain the extra trait bounds on the timestamp just to the methods that need it. Attempting to do so forced me to annotate every single method that transitively called the methods that lead to `Stash` calls and in the end there were more type annotations than the current version. If you have any ideas on how to reduce the noise I'm happy to try another approach.
* @benesch I'll like your eyes in particular on how the catalog migration is handled. Specifically, the migration code resolves this path `/mzdata/catalog/../controllers/storage/timestamp_bindings`. It seemed simpler than wiring up a separate top level path parameter just for this reason but if you think this is a smell I can wire up another parameter that is just the `/mzdata` path.

frankmcsherry commented:
> Attempting to do so forced me to annotate every single method that transitively called the methods that lead to Stash calls and in the end there were more type annotations than the current version.

I'm going to guess that those annotations were on the code that actually make the assumptions, though, vs the code that doesn't make the assumptions? We've been working on removing these constraints wherever possible, and having them creep back in because it makes other parts of the code easier is ... frustrating.

petrosagg commented:
> I'm going to guess that those annotations were on the code that actually make the assumptions, though, vs the code that doesn't make the assumptions? We've been working on removing these constraints wherever possible, and having them creep back in because it makes other parts of the code easier is ... frustrating.

The code that actually makes the assumptions is quite a bit larger than I thought infecting multiple methods of the storage controller and multiple methods of the compute controller. Basically any function that can transitively cause a storage compaction needs to have these bounds in place, and it turns out that the majority of the functions have some control flow that leads there. It's not about making the code easier, it's less noise overall. If you feel strongly about this I can put it on a function by function basis and just live with these bounds all over the place

benesch commented on src/dataflow-types/src/client/controller/storage.rs:
```diff
@@ -326,11 +496,25 @@ impl<'a, T: Timestamp + Lattice> StorageControllerMut<'a, T> {
+                        .map(|t| t.clone().try_into().expect("timestamp overflowed i64")),
+                );
+                ts_binding_stash.compact(since.borrow())?;
+                ts_binding_stash.consolidate()?;
```
Consolidating on every update might be expensive. Not sure! It matches what we currently do, perhaps, and so is maybe fine?

benesch commented on src/coord/src/catalog/storage.rs:
```diff
@@ -128,22 +148,80 @@ const MIGRATIONS: &[&str] = &[
+            .collect::<Result<Vec<_>, Error>>()?;
+
+        std::fs::create_dir_all(data_dir_path.join("controllers/storage/")).unwrap();
+        let stash = Stash::open(&data_dir_path.join("controllers/storage/timestamp_bindings"))
```
I was thinking that we'd have one `Stash` per layer, and namespace collection IDs as necessary. So `controllers/storage` would have `timestamp-bindings-1`, `timestamp-bindings-2`, etc. That leaves open the possibility of adding transactions to `Stash`, and then allowing each layer to transactionally update any of its collections. (Adding cross-collection transactions seems straightforward; cross-`Stash` transactions are far fetched though.)

benesch reviewed (approved):
> The Result wrapping stops at the coordinator where all the calls to the controllers are .unwrap()ed. This is what we were doing already and I didn't want to delay this PR with further error type refactoring.

This is great, thanks!

> It seemed simpler than wiring up a separate top level path parameter just for this reason but if you think this is a smell I can wire up another parameter that is just the /mzdata path.

It seems like you already went this direction? I left you some comments about dropping the `mzdata/catalog` argument now, though, since it seems you're passing through `mzdata` and could just construct `mzdata/catalog` on the fly.

petrosagg commented:
@benesch I addressed all your comments with the exception of consolidation. I'd like to see how that performs. One interesting thing to note is that this new implementation doesn't add new rows in the database as the upper ticks forwards because we're storing the bindings in differentialized form. Whereas the current main repeatedly stores the same statement over and over again so the data in SQLite should be significantly smaller for sources that don't produce data continuously

frankmcsherry commented on src/dataflow-types/src/client/controller/storage.rs:
```diff
@@ -267,11 +364,90 @@ impl<T: Timestamp + Lattice> StorageController for Controller<T> {
+            // together offsets and collapsing their timestamps. For example, if we were to compact
+            // thorugh timestamp 2 the collection would contain the following updates:
+            //
+            // (PartitionId::Kafka(1), 1, +7)  <-- 5 + 2 = 7
+            // (PartitionId::Kafka(2), 1, +17) <-- 10 + 7 = 17
```
Probably want the timestamps to be `2` here.

frankmcsherry commented on src/dataflow-types/src/client/controller/storage.rs:
```diff
@@ -267,11 +364,90 @@ impl<T: Timestamp + Lattice> StorageController for Controller<T> {
+            // (PartitionId::Kafka(2), 1, +17) <-- 10 + 7 = 17
+            ts_binding_stash.update_many(bindings.iter().cloned().flat_map(
+                |(pid, ts, offset)| {
+                    let prev_offset = collection_state
+                        .last_reported_ts_bindings
+                        .entry(pid.clone())
+                        .or_default();
+
+                    let ts = ts.try_into().expect("timestamp overflowed i64");
+                    let update = ((pid, ()), ts, offset.offset - prev_offset.offset);
+
```
The `prev_offset` update below maybe addresses this, but .. is this list sorted? It's all a bit confused, having these state updates applied in a `flat_map` that is an argument to `update_many`.

petrosagg replied:
I updated this section to have an explicit sort step and the calculations in a for loop. Having multiple updates for `pid` at the same timestamp will just create zero diff changes, which is fine.

mjibson commented on src/stash/src/lib.rs:
```diff
@@ -117,26 +118,36 @@ impl Stash {
+            }
+        };
+
+        tx.commit().unwrap();
```
Why `unwrap` instead of `?`?

petrosagg replied:
oops. Debugging leftover, I was trying to hunt down an error with a stacktrace. fixed

cjubb39 commented:
Any chance that re-enabling the tests disabled in #10942 passes CI now? :)

From `test/testdrive/kafka-exactly-once-sink.td`
